pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,283 @@
+		metrics.IngestorRequestsQueued.With(prometheus.Labels{"type": "trace", "queue_idx": fmt.Sprintf("%d", batchIdx)}).Dec()
+		batch.add(req)
+		if batch.addCounter == 1 {
+			// we reset timeout once we add first request into the batch
+			ticker.Reset(b.config.BatchTimeout)
+		}
+		if batch.isFull() {
+			metrics.IngestorBatchFlushTotal.With(prometheus.Labels{"type": "trace", "reason": "size"}).Inc()
+			batch = flushBatch(batch)
+		}
+	}
+	for {
+		select {
+		case item := <-b.in[batchIdx]:
+			processReq(item)
+		case <-ticker.C:
+			batcherSpan.AddEvent("Batch timeout reached")
+			metrics.IngestorBatchFlushTotal.With(prometheus.Labels{"type": "trace", "reason": "timeout"}).Inc()
+			if !batch.isEmpty() {
+				batch = flushBatch(batch)
+			}
+		case <-b.stop:
+			timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+			defer cancel()
+			close(b.in[batchIdx])
+			// shutting down. let's drain the requests
+		loop:
+			for {
+				select {
+				case item, ok := <-b.in[batchIdx]:
+					if !ok {
+						break loop
+					}
+					processReq(item)
+				case <-timeoutCtx.Done():
+					log.Warn("msg", "Forced batcher shutdown due to timeout. Some request might not be persisted.", "batcher", batchIdx)
+					break loop
+				}
+			}
+			if !batch.isEmpty() {
+				flushBatch(batch)
+			}
+			return
+		}
+	}
+}
+
+func (b *Batcher) Stop() {
+	b.once.Do(func() {
+		close(b.stop)
+		b.wg.Wait()
+		close(b.bufferedBatches)
+		b.batchWriter.stop()
+	})
+}
+
+// batchWriter writes batches using a writer.
+type batchWriter struct {
+	batches    chan Batch
+	numWriters int
+	writer     Writer
+	stopCh     chan struct{}
+	once       sync.Once
+	wg         sync.WaitGroup
+}
+
+func newBatchWriter(writers int, writer Writer, batches chan Batch) *batchWriter {
+	return &batchWriter{
+		writer:     writer,
+		numWriters: writers,
+		batches:    batches,
+		stopCh:     make(chan struct{}),
+	}
+}
+
+func (bw *batchWriter) run() {
+	for i := 0; i < bw.numWriters; i++ {
+		bw.wg.Add(1)
+		go func() {
+			defer bw.wg.Done()
+			for {
+				select {
+				case b, ok := <-bw.batches:
+					if !ok {
+						return
+					}
+					metrics.IngestorPendingBatches.With(prometheus.Labels{"type": "trace"}).Dec()
+					bw.flush(b)
+				case <-bw.stopCh:
+					timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+					defer cancel()
+					bw.drainBuffer(timeoutCtx)
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (bw *batchWriter) flush(b Batch) {
+	_, flushSpan := tracer.Default().Start(b.ctx, "flush-batch")
+	defer flushSpan.End()
+	err := bw.writer.InsertTraces(context.Background(), b.traces)
+	for _, req := range b.reqStatus {
+		req <- err
+	}
+
+}
+
+func (bw *batchWriter) drainBuffer(ctx context.Context) {
+	for {
+		select {
+		case b, ok := <-bw.batches:
+			if !ok {
+				return
+			}
+			bw.flush(b)
+		case <-ctx.Done():
+			log.Warn("msg", "Forced batchWriter shutdown due to timeout. Some batches might not be persisted.")
```

Comment:
again can we actually check if some batches weren't persisted before printing this msg